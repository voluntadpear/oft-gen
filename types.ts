export type TaskMetadata = {
  name: string;
  url: string | null;
};

export type Category = "PR" | "Issue" | "Support" | "Doc" | "Other";
export type Provider =
  | "GitHub"
  | "Google Docs"
  | "Zendesk"
  | "Notion"
  | "Slack";

export function assert(condition: boolean, msg?: string): asserts condition {
  if (!condition) {
    throw new Error(msg);
  }
}
