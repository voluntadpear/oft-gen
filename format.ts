import { TaskMetadata } from "./types.ts";

export function formatOFTMessage(tasks: TaskMetadata[]) {
  const template = `*Out for Today:*\n`;
  const bullets = tasks
    .map((task) => {
      const base = `•  ${task.name}`;
      if (task.url) {
        return `${base}\n\t•  [${prettifyHost(task.url)}](${task.url})`;
      }
      return base;
    })
    .join("\n");

  const output = `${template}${bullets}`;
  return output;
}

type Provider = "GitHub" | "Google Docs" | "Zendesk" | "Notion";
const originsMap: Record<string, Provider> = {
  "github.com": "GitHub",
  "google.com": "Google Docs",
  "zendesk.com": "Zendesk",
  "notion.so": "Notion",
};

function prettifyHost(url: string) {
  const parsedURL = new URL(url);
  const keys = Object.keys(originsMap);
  const match = keys.find((providerKey) =>
    parsedURL.origin.includes(providerKey)
  );
  if (!match) {
    return "Link";
  }

  const matchedValue = originsMap[match];

  if (matchedValue === "GitHub") {
    // Custom logic for both PRs and issues
    if (parsedURL.pathname.includes("/pull/")) {
      const prNumberRegex = /\/pull\/(?<prNumber>[0-9]+)/;
      const regexMatch = parsedURL.pathname.match(prNumberRegex);
      const prNumber = regexMatch?.groups?.prNumber;
      return `GitHub PR${prNumber ? ` #${prNumber}` : ""}`;
    } else if (parsedURL.pathname.includes("/issues")) {
      const issueRegex = /\/issues\/(?<issue>[0-9]+)/;
      const regexMatch = parsedURL.pathname.match(issueRegex);
      const issue = regexMatch?.groups?.issue;
      return `GitHub issue${issue ? ` #${issue}` : ""}`;
    }
    return "GitHub";
  }
  if (matchedValue === "Zendesk") {
    const ticketRegex = /\/tickets\/(?<ticket>[0-9]+)/;
    const regexMatch = parsedURL.pathname.match(ticketRegex);
    const ticket = regexMatch?.groups?.ticket;
    return ticket ? `Zendesk ticket #${ticket}` : "Zendesk";
  }

  return matchedValue;
}
