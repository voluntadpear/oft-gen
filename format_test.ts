import { assertSnapshot } from "https://deno.land/std@0.155.0/testing/snapshot.ts";

import { formatOFTMessage } from "./format.ts";
import { TaskMetadata } from "./types.ts";

Deno.test("format message", async (t) => {
  const tasks: TaskMetadata[] = [
    {
      name: "Task 1",
      url: null,
    },
    {
      name: "Task 2",
      url: "https://github.com/my-account/my-repo/pull/190",
    },
    {
      name: "Task 3",
      url: "https://docs.google.com/document/d/dfgokkjlsdf23453245blahblah/edit",
    },
    {
      name: "Task 4 with a longer description",
      url: "https://myorg.zendesk.com/agent/tickets/341",
    },
    {
      name: "Task 5",
      url: "https://myorg.zendesk.com/agent/filters/32311",
    },
    {
      name: "Task 6",
      url: "https://www.notion.so/abcdefoo",
    },
    {
      name: "Task 7",
      url: "https://github.com/orgs/myorg/projects/99",
    },
    {
      name: "Task 8",
      url: "https://github.com/myorg/myrepo/issues/991",
    },
  ];

  const output = formatOFTMessage(tasks);
  await assertSnapshot(t, output);
});
