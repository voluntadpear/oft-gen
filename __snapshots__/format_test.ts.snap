export const snapshot = {};

snapshot[`format message 1`] = `
"*Out for Today:*
•  ✅ Task 1
•  💡 Task 2 → [GitHub PR #190](https://github.com/my-account/my-repo/pull/190)
•  📄 Task 3 → [Google Docs](https://docs.google.com/document/d/dfgokkjlsdf23453245blahblah/edit)
•  ⚙️ Task 4 with a longer description → [Zendesk ticket #341](https://myorg.zendesk.com/agent/tickets/341)
•  ✅ Task 5 → [Zendesk](https://myorg.zendesk.com/agent/filters/32311)
•  📄 Task 6 → [Notion](https://www.notion.so/abcdefoo)
•  ✅ Task 7 → [GitHub](https://github.com/orgs/myorg/projects/99)
•  🔦 Task 8 → [GitHub issue #991](https://github.com/myorg/myrepo/issues/991)"
`;
